src/main/java/org/apache/cassandra/sidecar/routes/FileStreamHandler.java

```diff
@@ -74,40 +103,114 @@ protected String extractParamsOrThrow(RoutingContext context)
         return context.get(FILE_PATH_CONTEXT_KEY);
     }
 
+    @Override
+    protected void processFailure(Throwable cause,
+                                  RoutingContext context,
+                                  String host,
+                                  SocketAddress remoteAddress,
+                                  String localFile)
+    {
+        IOException fileNotFoundException = ThrowableUtils.getCause(cause, NoSuchFileException.class);
+
+        if (fileNotFoundException == null)
+        {
+            // FileNotFoundException comes from the stream method
+            fileNotFoundException = ThrowableUtils.getCause(cause, FileNotFoundException.class);
+        }
+
+        if (fileNotFoundException != null)
+        {
+            logger.error("The requested file '{}' does not exist", localFile);
+            context.fail(wrapHttpException(NOT_FOUND, "The requested file does not exist"));
+            return;
+        }
+
+        super.processFailure(cause, context, host, remoteAddress, localFile);
+    }
+
     /**
-     * Ensures that the file exists and is a non-empty regular file
+     * Retrieves the Future for the validation operation. If the cache was initialized, it will retrieve the
+     * Future from the cache.
      *
      * @param fs        The underlying filesystem
      * @param localFile The path the file in the filesystem
-     * @param exists    Whether the file exists or not
      * @return a succeeded future with the {@link FileProps}, or a failed future if the file does not exist;
      * is not a regular file; or if the file is empty
      */
-    private Future<FileProps> ensureValidFile(FileSystem fs, String localFile, Boolean exists)
+    protected Future<FileProps> ensureValidFile(FileSystem fs, String localFile)
     {
-        if (!exists)
+        if (filePropsCache != null)
         {
-            logger.error("The requested file '{}' does not exist", localFile);
-            return Future.failedFuture(wrapHttpException(NOT_FOUND, "The requested file does not exist"));
+            FileProps fileProps = filePropsCache.getIfPresent(localFile);
+
+            if (fileProps != null)
+            {
+                return Future.succeededFuture(fileProps);
+            }
+
+            return ensureValidFileNonCached(fs).apply(localFile)
```

Comment:
I am not convinced on the usefulness of the cache.

1. When the cache is update to date, the cache helps to save 1 look up of the file attributes. However, during `sendFile`, there are more I/O operations. 
2. When the cache is stale, the file does not exist. It is going to acquire permit and try to `sendFile` falsely, which could undesirably increase CPU usage.

Reply:
Removed this cache